app/controllers/phases_controller.rb

```diff
@@ -1,44 +1,72 @@
+# frozen_string_literal: true
+
 class PhasesController < ApplicationController
-  before_action :set_phase, only: %i[show edit update destroy]
+  before_action :set_phase, only: %i[show edit update engineer destroy complete]
   before_action :set_project_lead
+  before_action :set_engineer, only: :show
 
   def index
     @phases = @project_lead.phases
   end
 
-  def show; end
+  def show
+    @users = User.joins(:roles).where('roles.name =? ', 2)
+  end
 
   def new
     @phase = @project_lead.phases.new
+    authorize @phase
   end
 
-  def edit; end
+  def edit
+    authorize @phase
+  end
 
   def create
     @phase = @project_lead.phases.new(phase_params)
+    authorize @phase
     return render :new, notice: 'Phase not Saved !' unless @phase.save
 
     redirect_to project_lead_phases_path, notice: 'Phase was successfully created.'
   end
 
+  def engineer
+    engineer = User.find(params[:engineer][:user_id])
+    if @phase.users.exists?(engineer.id)
+      flash[:alert] = 'Already Assigned'
+    else
+      @phase.users.append(engineer)
+      flash[:notice] = 'Engineer Assigned.'
+    end
+    redirect_to project_lead_phase_url(@phase.project_lead_id)
```

Comment:
This type of validation should be in model.

Reply:
They are performing some action that's why they are used in controllers

Reply:
They are performing some action on button click so they are in controller. Engineer is assigning engineers to phase from the drop down.